Sources/GRPC/GRPCWebToHTTP2ServerCodec.swift

```diff
@@ -0,0 +1,482 @@
+      /// stream.
+      case open(OpenState)
+
+      /// Closed; the response stream (and therefore the request stream) has been closed.
+      case closed
+
+      /// Not a real state.
+      case _modifying
+    }
+
+    fileprivate struct OpenState {
+      /// A `ByteBuffer` containing the base64 encoded bytes of the request stream if gRPC Web Text
+      /// is being used, `nil` otherwise.
+      var requestBuffer: ByteBuffer?
+
+      /// A `CircularBuffer` holding any response messages if gRPC Web Text is being used, `nil`
+      /// otherwise.
+      var responseBuffer: CircularBuffer<ByteBuffer>?
+
+      /// True if the end of the request stream has been received.
+      var requestEndSeen: Bool
+
+      /// True if the response headers have been sent.
+      var responseHeadersSent: Bool
+
+      init(isTextEncoded: Bool, allocator: ByteBufferAllocator) {
+        self.requestEndSeen = false
+        self.responseHeadersSent = false
+
+        if isTextEncoded {
+          self.requestBuffer = allocator.buffer(capacity: 0)
+          self.responseBuffer = CircularBuffer()
+        } else {
+          self.requestBuffer = nil
+          self.responseBuffer = nil
+        }
+      }
+    }
+  }
+}
+
+extension GRPCWebToHTTP2ServerCodec.StateMachine.State {
+  fileprivate mutating func processInbound(
+    serverRequestPart: HTTPServerRequestPart,
+    allocator: ByteBufferAllocator
+  ) -> GRPCWebToHTTP2ServerCodec.StateMachine.Action {
+    switch serverRequestPart {
+    case let .head(head):
+      return self.processRequestHead(head, allocator: allocator)
+    case var .body(buffer):
+      return self.processRequestBody(&buffer)
+    case .end:
+      return self.processRequestEnd(allocator: allocator)
+    }
+  }
+
+  fileprivate mutating func processOutbound(
+    framePayload: HTTP2Frame.FramePayload,
+    promise: EventLoopPromise<Void>?,
+    allocator: ByteBufferAllocator
+  ) -> GRPCWebToHTTP2ServerCodec.StateMachine.Action {
+    switch framePayload {
+    case let .headers(payload):
+      return self.processResponseHeaders(payload, promise: promise, allocator: allocator)
+
+    case let .data(payload):
+      return self.processResponseData(payload, promise: promise)
+
+    case .priority,
+         .rstStream,
+         .settings,
+         .pushPromise,
+         .ping,
+         .goAway,
+         .windowUpdate,
+         .alternativeService,
+         .origin:
+      preconditionFailure("Unsupported frame payload")
+    }
+  }
+}
+
+// MARK: - Inbound
+
+extension GRPCWebToHTTP2ServerCodec.StateMachine.State {
+  private mutating func processRequestHead(
+    _ head: HTTPRequestHead,
+    allocator: ByteBufferAllocator
+  ) -> GRPCWebToHTTP2ServerCodec.StateMachine.Action {
+    switch self {
+    case .idle:
+      var headers = HPACKHeaders(httpHeaders: head.headers, normalizeHTTPHeaders: true)
+      headers.add(name: ":path", value: head.uri)
+      headers.add(name: ":method", value: head.method.rawValue)
```

Comment:
These pseudo-headers are surely not going to be in the right place in the block, right? And I think we're missing two: `:authority` and `:scheme` are both expected as well.

Reply:
I didn't think the order mattered given these will only be read by the HTTP/2 to gRPC codec? Easy enough to change though.

I wasn't sure about including `:authority` and `:scheme`; they're also not required but I suppose they _might_ be expected by a user in e.g. an interceptor. WDYT?

Reply:
I'm inclined to suggest that we should do the right thing here. It reduces the risk that we might accidentally introduce a component that doesn't expect this and get issues down the road. The same applies to adding the extra headers.

Reply:
That seems wise. Done!